### ISSUE TITLE
Accessibility: List widget keyboard access to actions

### ISSUE THREAD

**Issue by cleidigh** [Microsoft/vscode#24119]
All versions:

Both the new SCM and revamped extension viewlits use the new list wg.
If one is using keyboard navigation each of these require several extra steps and duplicated tab stop
or procedure steps. 

Basic example for extensions:

- user has copied a marketplace link or entered a search term
- user has to Tab into extension list
- press X number of Down keys
- press Enter - entry gets focus - default entry action opens extension description in editor
- ActionBar actions for the selective list item cannot be accessed unless user Tab stops
all the way from the beginning back to the install button (or gear icon) 
- knowing where their current tab stop is can be difficult as it might be out of view
and for current contrast level , color and width of the Tab stop highlight is minimal

With new SCM:
- default keyboard action on Enter to select also open file item 
- in order to get the per file item actions the user has to go back to the widget a second time
and then use the available Tab stops to get to the actions

Possible joint solutions:
- Selecting a list item with the keyboard should make the next Tab stop to the action of the item actions
- don't separate focus from selection by keyboard like the old git source widget (actions available on selection)
- use Control Enter to select and activate actions availability or another key
- add user settings item that makes selection and focus overlapped for keyboard only users benefit

I see accessibility items as main part of April milestones.

I would do a PR and work on this if accepted...
Thanks

**Comment by joaomoreno** [comment]
@cleidigh Sure, let's try a PR.

**Comment by cleidigh** [comment]
@joaomoreno 
great
Protocol question - do you guys prefer to discuss first in the issue thread or have me open a PR
and discuss there?

Either way I will start jumping into the list code to get a better handle on it.

**Comment by joaomoreno** [comment]
PR discussion would be great.

**Comment by cleidigh** [comment]
@joaomoreno 
bit of an update
- have done a lot of work analyzing the list in action bar code
- almost have a proof of concept done showing some changes for the extension 's viewlet
- I think this will help drive the conversation
- it's mostly about tabindex but other navigation keys sometimes also
- I hope to push some code up to my repository for demo and conversation maybe tomorrow

**Comment by cleidigh** [comment]
@joaomoreno 
Awesome, this makes the SCM keyboard workflow much better !
Thanks.

**Comment by roblourens** [comment]
I see the improvement for the git viewlet, but I can't figure out how to tab from a selected extension to the 'install' button on the list item. Is that part of this fix?

**Comment by cleidigh** [comment]
@roblourens 
I did a PR on this original issue. @joaomoreno Suggested we separate the issues and he subsequently
addressed SCM with a slightly different approach.
Here is the original canceled PR:
https://github.com/Microsoft/vscode/pull/24628/files
The approach was the following:
- Add a constructor option for the action bars to allow the parent to control its tab index
- For the extension sidebar I added the ability for the parent to toggle the action bars tab index
  on when the line was focused
- This allows you to Tab directly to the buttons and action bar icons when you have the line focused
- The extension sidebar is a separate case since as you know the icons are always visible

The PR is rather old and I've learned a lot since then but I think the approach is pretty decent

I would like to redo the PR for August if you guys are good with it.

**Comment by roblourens** [comment]
I see, if this issue just covers the SCM viewlet I'll call it verified.

Is there another issue already for the extensions viewlet?

**Comment by cleidigh** [comment]
@roblourens 
I will create a new issue for the extensions.
